js/lib/AlphabeticIndex.js

```diff
@@ -221,128 +386,322 @@ AlphabeticIndex.prototype.addLabels = function(labels, start) {
- * 
+ *
  * @returns {Object} a hash of all buckets and elements
  * as per the description above.
  */
 AlphabeticIndex.prototype.getAllBuckets = function() {
+    var underflowIndex = -1, overflowIndex = -1;
+    var mixedScriptEndIndex = -1;
+    var count = 0;
+    var temp;
+    var i;
+
+    var tempArr = [];
+    var tempIndex = [];
+    var tempBucket = {};
+    var itemIndex;
+
+    for (i=0; i < this.index.length; i++) {
+        tempArr.push(this.index[i].label);
+    }
+
+    tempArr.sort(this.collationObj.compare());
+
+    for (i=0; i < tempArr.length; i++) {
+        tempBucket={};
+        tempBucket.label = tempArr[i];
+        itemIndex = this._getLabelIndex(tempArr[i]);
+
+        this.index[itemIndex].elements.sort(this.collationObj.compare());
+        tempBucket.elements = this.index[itemIndex].elements;
+        tempIndex[i] = tempBucket;
+        tempBucket={};
+    }
+
+    this.index = tempIndex;
+
+    for (i=0; i < this.index.length; i++) {
+        if (this.inherit &&
+            this.mixedCollationMap[0].indexUnits.indexOf(this.index[i].label) == -1) {
+            mixedScriptEndIndex = i;
+            count++;
+        }
+    }
+
+    if (this.inherit && count > 0) {
+        temp = this.index.splice((mixedScriptEndIndex - count) +1 , count);
+        this.index = this.index.concat(temp);
+    }
+
+    for (i=0; i < this.index.length; i++) {
+        if (this.index[i].label == this.underflowLabel) {
+            underflowIndex = i
+            break;
+        }
+    }
+
+    if (underflowIndex > 0) {
+        temp = this.index.splice(underflowIndex,1)[0];
+        this.index.unshift(temp);
+    }
+
+    for (i=0; i < this.index.length; i++) {
+        if (this.index[i].label == this.overflowLabel) {
+            overflowIndex = i
+            break;
+        }
+    }
+
+    if (overflowIndex > -1) {
+        temp = this.index.splice(overflowIndex,1)[0];
+        this.index.push(temp);
+    }
+
+    return this.index;
 };
 
 /**
- * Return the label of the bucket for a given element. This 
+ * Return the label of the bucket for a given element. This
  * follows the rules set up when the index was instantiated to
  * find the bucket into which the element would go if it
  * were added to this index. The element is not added to
  * the index, however. (See addElement for that.)
- * 
- * @param {String} element the element to check
- * @returns {String} the label for the bucket for this element
+ *
+ * @param {string|undefined} element the element to check
+ * @returns {string|undefined} the label for the bucket for this element
  */
 AlphabeticIndex.prototype.getBucket = function(element) {
+    var label;
+    var firstChar;
+    var collationValue;
+    var charNum, firstBoundaryChar, endBoundaryChar, firstCharNum, endCharNum;
+    var source;
+    var validMapNum = -1;
+    
+    if (!element) {
+        return undefined;
+    }
+
+    source = new NormString(element);
+    firstChar = source.str[0];
```

Comment:
What you have on these two lines is equivalent to

```
firstChar = element[0];
```

... because the code doesn't do anything with the string once it is contained within the NormString instance. Specifically, it does not perform any normalization on the string "element" before looking at the first character. Why create a NormString instance on a string if you are not going to normalize that string?

Instead, this should call 

```
source = new NormString(element);
source.normalize("nfc"); 
var it = source.charIterator();
firstChar = it.next();
```

To put the string into a standard form first and take the first character of that normalized string.